### PR TITLE
Fix dev service always restarting on named datasource configuration change

### PR DESCRIPTION
**Pull request by Malandril**
When a named datasource is used, quarkus tries to check if a some of the `cachedProperties` have changed to know if it needs to restart the datasource, the `cachedProperties` stores the properties with and without quotes.
This causes the change check to always be triggered, because when it will compare with the actual microprofile property only one version will exist, either the quoted version, or the unquoted version.
For example with the given property:
```properties
quarkus.datasource."named".db-kind=mariadb
```

`cachedProperties` will cache the following entries as set by `setDataSourceProperties`:
```properties
quarkus.datasource."named".db-kind=mariadb
quarkus.datasource.named.db-kind=mariadb
```

And when it will try to check the difference only the `quarkus.datasource."named".db-kind=mariadb` will be present and so the restart will be triggered.

This PR uses gets a map of the datasource configs which will be used as cache to verify if values have changed.

The drawback of this is that any change in the `quarkus.datasource` properties triggers a restart, which i think is still better than having a non consistent behaviour because of quotes.

[Bug Reproducer here](https://github.com/Malandril/quarkus-db-restart) that i used to test this PR.

### CODE REVIEW

**Review by gsmet — COMMENTED** [review]
Thanks for the PR! That's indeed an annoying issue.

TBH, I'm not entirely sure why we store the two versions in `cachedProperties`. Maybe to avoid triggering a reload if only the quoting changed. But yeah, breaking the common case for it doesn't make a lot of sense.

I wonder if we could find a better way to do that as it looks like a (useful!) hack on top a hack.

/cc @radcortez @yrodiere

**Comment by radcortez** [comment]
I'm not sure why we need to have both versions,but apparently there was another issue: See https://github.com/quarkusio/quarkus/pull/26815.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 1]
```diff
@@ -50,7 +51,8 @@ public class DevServicesDatasourceProcessor {
 
     // list of devservices properties we should not check for restart
     // see issue #30390
-    private static final Set<String> EXCLUDED_PROPERTIES = Set.of("quarkus.datasource.devservices.enabled");
+    private static final Set<String> EXCLUDED_PROPERTIES = Set.of("quarkus.datasource.devservices.enabled",
+            "quarkus.datasource.devservices.reuse");
```
Comment (yrodiere) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:55]: This looks unrelated? Please at least move to a separate commit and explain the rationale in the PR description + commit message.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 2]
```diff
@@ -77,13 +79,17 @@ DevServicesDatasourceResultBuildItem launchDatabases(
...
-                            trim(ConfigProvider.getConfig().getOptionalValue(entry.getKey(), String.class).orElse(null)))) {
-                        restartRequired = true;
-                        break;
-                    }
+            for (Map.Entry<String, String> entry : cachedProperties.entrySet()) {
+                String value = ConfigProvider.getConfig()
+                        .getOptionalValue(entry.getKey(), String.class)
+                        .orElse(null);
+                if (value == null) {
+                    value = getValueWithOrWithoutQuotes(entry.getKey(),
```
Comment (yrodiere) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:87]: I'm not blocking the merge; if it works let's merge.

But! Ideally we would instead:

1. (maybe) make sure to always normalize keys (use quotes) before putting them in `cachedProperties`
2. make sure that `getOptionalValue` does not differentiate between quoted and non-quoted syntax

@radcortez any opinion/insight about item 2? It would have made my fix in #26815 unnecessary.

Comment (radcortez) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:87]: When calling a `get` in the config system with the full property name, we require an exact match because we cannot tell if the quotes should be omitted or not. In most cases, we can't even control that because the source provides the value (and all sources do exact matches).

Alternatively, we could look the configuration as a `Map`, and in that case, the config system is smart enough to remove the quotes from the path key segments. You could do a `SmallRyeConfig#getOptionalValues("quarkus.datasource", String.class, String.class)`.

Comment (Malandril) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: Thanks for the feedback i pushed a new commit using `SmallRyeConfig#getOptionalValues` to create the cache which simplies this a lot.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 3]
```diff
@@ -77,20 +78,21 @@ DevServicesDatasourceResultBuildItem launchDatabases(
...
+                if (!Objects.equals(entry.getValue(), trim(newValue))) {
+                    restartRequired = true;
+                    break;
                 }
             }
             if (!restartRequired) {
-                //devservices properties may have been added
-                for (var name : ConfigProvider.getConfig().getPropertyNames()) {
-                    if (name.startsWith("quarkus.datasource.") && name.contains(".devservices.")
-                            && !cachedProperties.containsKey(name) && !EXCLUDED_PROPERTIES.contains(name)) {
```
Comment (Malandril) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:93]: I don't think the excluded properties are necessary anymore, as we get all the datasource properties with `SmallRyeConfig#getOptionalValues`

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 4]
```diff
@@ -172,7 +174,10 @@ public void run() {
             closeBuildItem.addCloseTask(closeTask, true);
         }
         databases = runningDevServices;
-        cachedProperties = propertiesMap;
```
Comment (Malandril) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:175]: I think that we could even remove the _propertiesMap_ entirely, and juste keep devDebProperties

**Comment by Malandril** [comment]
The last commit actually seems to break the build on the pipelines.

In the `jpa-postgresql` integration test.
A  `java.util.NoSuchElementException: SRCFG00011: Could not expand value postgres.url in property quarkus.datasource.jdbc.url`

As apparently when creating the cached properties, the `postgres.url` is not available.
This does not happen when i run the test myself.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java (hunk 4 above)
Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:175]: While I agree your solution looks cleaner, we need to think about which properties need to trigger a restart.

Previously, only a change in properties populated in `propertiesMap` would trigger a restart. Things like db kind, port, username, dev service properties, etc. Things that actually affect the dev service.

After your changes, any property starting with `quarkus.datasource` will trigger a restart of the container... including things like `quarkus.datasource.metrics.enabled` or `quarkus.datasource.jdbc.driver` or `quarkus.datasource.jdbc.min-size` ... which are clearly client-side?!


If you want to do things this way, I think you need to filter the maps obtained through `getOptionalValues` to only keep the properties that would actually matter for dev services (i.e. the ones we used to put in `propertiesMap`).

**Review by yrodiere — CHANGES_REQUESTED** [review]
`SmallRyeConfig#getOptionalValues` seems promising, but I think we got carried away here... see comments below.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java (hunk 3 above)
Comment (radcortez) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:93]: > We may need to adjust the content of `EXCLUDED_PROPERTIES`, though, because I suppose property keys are now relative to `quarkus.datasource.`?

Yes, the path to the `Map` is not included in the keys, so the keys are everything after `quarkus.datasource`.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 5]
```diff
@@ -172,14 +189,47 @@ public void run() {
...
+    private static Map<String, String> getRestartTriggeringPropertiesValues() {
+        Map<String, String> datasourceProperties;
+        try {
+            datasourceProperties = ConfigProvider.getConfig()
+                    .unwrap(SmallRyeConfig.class)
+                    .getOptionalValues("quarkus.datasource",
+                            String.class,
+                            String.class)
+                    .orElse(Map.of());
+        } catch (NoSuchElementException e) {
```
Comment (Malandril) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:213]: This happens in integration tests _integration-tests/jpa-postgresql_ the maven property `postgresq.url` is not resolved unknown at the time the datasource properties are queried, and the test fails.

I don't know if this is a normal an expected behaviour, so i put this catch. We can remove it if itdoesn't make sense

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 6]
```diff
@@ -50,7 +52,25 @@ public class DevServicesDatasourceProcessor {
...
+            "db-version",
+            "devservices.command",
+            "devservices.container-env.",
+            "devservices.container-properties.",
+            "devservices.db-name",
+            "devservices.image-name",
+            "devservices.init-script-path",
+            "devservices.port",
+            "devservices.properties.",
+            "devservices.reuse",
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:70]: I think it makes sense to restart when a value *may* impact how we start the container.
So yes, anything used in `DevServicesDatasourceContainerConfig` (among others) should trigger a restart: it's passed to testcontainers when starting the container.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java (hunk 5 above)
Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:213]: This is concerning... and no, I don't think catching and ignoring the exception is acceptable here.

@radcortez do you think this could be related to the class `io.quarkus.it.jpa.postgresql.OverrideJdbcUrlBuildTimeConfigSource` you added to that module?

**Review by yrodiere — COMMENTED** [review]
Thanks, the approach looks good. Some details need our attention though; see comments below.

In particular we'll need @radcortez opinion on that exception.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java (hunk 5 above)
Comment (radcortez) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: Yes it is.

For now, I've just pushed an update to disable expression expansion in that piece of code, but this is a bug in SmallRye Config. Any call to `getOptional*` should not give an error if the expression cannot be expanded.

This also means that we cannot reliably monitor configuration names that target an expression, at least with the current implementation. Because an expression may only be available at runtime, the cache would have to be built in a runtime recorder.

Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: It seems your fix allowed one test to pass, but uncovered other errors...

> This also means that we cannot reliably monitor configuration names that target an expression, at least with the current implementation. Because an expression may only be available at runtime, the cache would have to be built in a runtime recorder.

This code starts dev services and needs to happen in build steps, though.
Are you saying we cannot use `getOptional*` in this PR and must find another way?

Comment (Malandril) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: Is the solution, with the catch even if not covering all cases, is it not already better than the current behaviour of Always restarting ?
Because in the worst case of a thrown error, it will simply mean that the dev services will be restarted each time. i could not reproduce it ourside of the test by providing runtime peoperties, and variable substitution.

Alternatively, the solution on the first commit on the PR works without those issues, although with a clunky solution.

Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: > Is the solution, with the catch even if not covering all cases, is it not already better than the current behaviour of Always restarting ?

It is not.

The current solution does not "always restart" incorrectly. It restarts incorrectly in a specific case you reported, when changing quotes.

As far as I know, the current solution does not restart "all the time" when a property value contains `${postgres.url}`. The new solution would.

I'm open to reverting to your original solution before your switched to `getOptionalValues()`, but my comment would still stand: 
https://github.com/quarkusio/quarkus/pull/41736#discussion_r1668653666

A third solution would be to extract the data from `DataSourceBuildTimeConfig` directly. But that solution is obvious, so much simpler, that I'm pretty sure there was a good reason not to do that.

In any case... let's wait for @radcortez 's insights before doing anything else.

Comment (Malandril) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: > The current solution does not "always restart" incorrectly. It restarts incorrectly in a specific case you reported, when changing quotes.
> 
Just to be precise it Always restarts, even with no changes, if there is at least one named datasource, quoted or not.

Comment (radcortez) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: > This code starts dev services and needs to happen in build steps, though.
> Are you saying we cannot use `getOptional*` in this PR and must find another way?

You can. The expression expansion issue that I brought up also happened with the old implementation, but I guess it is a niche case.

To summarize, we cannot safely expand runtime expressions during build time (where we populate the cache), because the configuration source may not be available at this point. So if we cache `quarkus.datasource.jdbc.url` to `${postgres.url}`, the user will change the value of `postgres.url`, which would not trigger the restart. I guess we could add such expressions to the monitor list.

Comment (radcortez) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: > A third solution would be to extract the data from `DataSourceBuildTimeConfig` directly. But that solution is obvious, so much simpler, that I'm pretty sure there was a good reason not to do that.

Probably because mappings didn't support `equals` and `hashcode` which they do now. So you could either do it for the entire object, or selected pieces.

Comment (Malandril) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: > Probably because mappings didn't support `equals` and `hashcode` which they do now. So you could either do it for the entire object, or selected pieces.

So we can simply call `equals` between two `DataSourceBuildTimeConfig` ?

Or do we still need a map with something like that we will compare ? :

```java
Map<String, Object> res = new HashMap<>();
for (var datasource : dataSourcesBuildTimeConfig.dataSources().entrySet()) {
    String name = datasource.getKey();
    DataSourceBuildTimeConfig config = datasource.getValue();
    res.put(name + ".db-kind", config.dbKind());
    res.put(name + ".db-version", config.dbVersion());
    res.put(name + ".devservices.command", config.devservices().command());
    res.put(name + ".devservices.container-env", config.devservices().containerEnv());
    res.put(name + ".devservices.container-properties.", config.devservices().containerProperties());
    res.put(name + ".devservices.db-name", config.devservices().dbName());
    res.put(name + ".devservices.enabled", config.devservices().enabled());
    res.put(name + ".devservices.image-name", config.devservices().imageName());
    res.put(name + ".devservices.init-script-path", config.devservices().initScriptPath());
    res.put(name + ".devservices.password", config.devservices().password());
    res.put(name + ".devservices.port", config.devservices().port());
    res.put(name + ".devservices.properties", config.devservices().properties());
    res.put(name + ".devservices.reuse", config.devservices().reuse());
    res.put(name + ".devservices.username", config.devservices().username());
    res.put(name + ".devservices.volumes", config.devservices().volumes());
}
```

Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: > So we can simply call equals between two DataSourceBuildTimeConfig ?

No, because then you'd be restarting too often, e.g. when `DataSourceBuildTimeConfig#healthExclude` changes.

> Or do we still need a map with something like that we will compare ? :

That's an option... or you hold the information you are interested in, and only that information, in something like this:

```
private record DevServiceConfigCache(
    Optional<String> dbKind,
    Optional<String> dbVersion,
    DevServicesBuildTimeConfig devservices
)
```

Comment (yrodiere) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: Actually scratch that, since apparently we don't want to restart when `quarkus.datasource.devservices.enabled` changes... your solution with a Map is better, more flexible. And more backportable.

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java [hunk 7]
```diff
@@ -42,40 +39,16 @@
...
-            "devservices.container-properties.",
-            "devservices.db-name",
-            "devservices.image-name",
-            "devservices.init-script-path",
-            "devservices.port",
-            "devservices.properties.",
-            "devservices.reuse",
-            "devservices.volumes.",
-            "password",
-            "username");
```
Comment (yrodiere) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java]: I think you'll need to get `quarkus.datasource[."name"].user` and `quarkus.datasource[."name"].password` as well, given a change to these properties used to trigger a restart... and for good reason: they are used as a default for devservice user/password:

https://github.com/quarkusio/quarkus/blob/e39d84e14516ebe006e0c31531511b8b496823ad/extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java#L289-L292

To avoid problems with quoted/unquoted, maybe you could get the values using the exact same code used above (`getFirstOptionalValue`)?

**Review by yrodiere — COMMENTED** [review]
Thanks a lot for your patience, I know it's been a rough ride but it's looking good in the end.

I think the last remaining piece is handling of two runtime properties... see below.

**Review by yrodiere — APPROVED** [review]

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
> Thanks for the review and the help @yrodiere @radcortez

No. Thank you for your patience :)

**Comment by yrodiere** [comment]
Merged, thanks a lot!

**Comment by yrodiere** [comment]
@gsmet Marking for backport on 3.8 as I *think* the final implementation would apply cleanly, and the problem looks like something that would impair development significantly.
